src/lib.rs

```diff
@@ -307,24 +308,65 @@ fn _encode(input: &[u8]) -> String {
 /// ```
 #[inline(always)]
 pub fn encode_to_str<R: AsRef<[u8]>>(input: R) -> String {
-    _encode(input.as_ref())
+    let buffer = encode_to_buffer(input.as_ref());
+    match String::from_utf8_lossy(&buffer) {
+        std::borrow::Cow::Borrowed(borrowed) => {
+            drop(borrowed);
+            // given that the vec could represent the string immutably
+            // the vec must be utf8 encodable. this next line cannot panic.
+            String::from_utf8(buffer).expect("the buffer must be utf8 encodable")
+        }
+        // note: all quoted-printable encoded strings should be utf8 encodable
+        // this case is here for completion but it should never occur in practice
+        std::borrow::Cow::Owned(owned) => owned,
+    }
```

Comment:
This bit seems to do the utf-8 decoding twice which seems unnecessarily inefficient. Really you can just do the `String::from_utf8(...)` bit and that should be equivalent. If the buffer contains non-utf8 (or even non-ascii) data then panicking is appropriate since it means there's a bug in this library.

I'm slightly concerned that we're trading off one thing for another here - adding the `Write` ability allows saving memory if you're using that API, but this is going to make `encode_to_str` unconditionally slower since it now has a utf-8 decode when it didn't before.